fendermint/vm/interpreter/src/chain.rs

```diff
@@ -91,30 +103,52 @@ where
+            msgs.push(ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                height: proposal.height as ChainEpoch,
+                block_hash: proposal.block_hash,
+            })))
+        }
+
         Ok(msgs)
     }
 
     /// Perform finality checks on top-down transactions and availability checks on bottom-up transactions.
-    async fn process(&self, pool: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
+    async fn process(&self, state: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
         for msg in msgs {
-            if let ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) = msg {
-                let item = CheckpointPoolItem::BottomUp(msg);
-
-                // We can just look in memory because when we start the application, we should retrieve any
-                // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
-                // We don't have to validate the checkpoint here, because
-                // 1) we validated it when it was relayed, and
-                // 2) if a validator proposes something invalid, we can make them pay during execution.
-                let is_resolved = atomically(|| match pool.get_status(&item)? {
-                    None => Ok(false),
-                    Some(status) => status.is_resolved(),
-                })
-                .await;
-
-                if !is_resolved {
-                    return Ok(false);
+            match msg {
+                ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) => {
+                    let item = CheckpointPoolItem::BottomUp(msg);
+
+                    // We can just look in memory because when we start the application, we should retrieve any
+                    // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
+                    // We don't have to validate the checkpoint here, because
+                    // 1) we validated it when it was relayed, and
+                    // 2) if a validator proposes something invalid, we can make them pay during execution.
+                    let is_resolved = atomically(|| match state.0.get_status(&item)? {
+                        None => Ok(false),
+                        Some(status) => status.is_resolved(),
+                    })
+                    .await;
+
+                    if !is_resolved {
+                        return Ok(false);
+                    }
                 }
-            }
+                ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                    height,
+                    block_hash,
+                })) => {
+                    let prop = IPCParentFinality {
+                        height: height as u64,
+                        block_hash,
+                    };
+                    return Ok(atomically(|| state.1.check_proposal(&prop)).await);
```

Comment:
```suggestion
                    let is_final = atomically(|| state.1.check_proposal(&prop)).await
                    if !is_final {
                      return Ok(false);
                    }
```

This seems to have regressed. Remember that we have lots of messages to check and this is just one.